File: src/2.两数相加.ts
/*
 * @lc app=leetcode.cn id=2 lang=typescript
 *
 * [2] 两数相加
 */

// @lc code=start
// Definition for singly-linked list.
// class ListNode {
//     val: number
//     next: ListNode | null
//     constructor(val?: number, next?: ListNode | null) {
//         this.val = (val===undefined ? 0 : val)
//         this.next = (next===undefined ? null : next)
//     }
// }

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {

    type listnode = ListNode | null

    let head:listnode = null, tail:listnode = null

    let carry = 0

    while(l1 || l2){

        const n1 = l1 ? l1.val : 0
        const n2 = l2 ? l2.val : 0

        const sum = n1 + n2 + carry

        if(!head){
            head = tail = new ListNode(sum % 10)
        } else if(tail) {
            tail.next = new ListNode(sum % 10)
            tail = tail.next
        }
        carry = Math.floor( sum / 10)

        if(l1) {
            l1 = l1.next
        }

        if(l2) {
            l2 = l2.next
        }

    }
    if(carry && tail) {
        tail.next = new ListNode(carry)
    }

    return head
};
// @lc code=end

File: src/2.两数相加.go
/*
 * @lc app=leetcode.cn id=2 lang=golang
 *
 * [2] 两数相加
 */
package leetcode

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var (
		dumbhead = &ListNode{
			Val: -1,
			Next: nil,
		}
		l1cur = l1
		l2cur = l2
		carry = 0
		dumbtail = dumbhead
	)
	
	for l1cur != nil || l2cur != nil {
		n1, n2 := 0,0
		if l1cur != nil {
			n1 = l1cur.Val
			l1cur = l1cur.Next
		}

		if l2cur != nil {
			n2 = l2cur.Val
			l2cur = l2cur.Next
		}
		
		sum := (n1+n2+ carry) % 10
		carry = (n1+n2+ carry) / 10
		dumbtail.Next = &ListNode{
			Val: sum,
		}
		dumbtail = dumbtail.Next
	}

	if carry > 0 {
		dumbtail.Next = &ListNode{
			Val: carry,
		}
	}

	return dumbhead.Next
}
// @lc code=end

